### PR TITLE
x86: Don't enable interrupt during exception handling when MMU is enabled

### PR DESCRIPTION
**Pull request by kwd-doodling**
Previously with interrupt enabled, double fault occurs if the code interrupting page fault handler triggered another page fault. And also disable code wrapped with CONFIG_LAZY_FPU_SHARING, which becomes unnecessary with interrupt disabled.

### CODE REVIEW

**Review by teburd — APPROVED** [review]

File: arch/x86/core/ia32/excstub.S [hunk 1]
```diff
@@ -130,6 +130,7 @@ SECTION_FUNC(PINNED_TEXT, _exception_enter)
 #endif
 	/* ESP is pointing to the ESF at this point */
 
+#ifndef CONFIG_X86_MMU
```
Comment (edersondisouza) [inline arch/x86/core/ia32/excstub.S:133]: Isn't this a bit too heavy handed? It seems that interrupts should not be enabled only if the MMU is handling an exception. Not re-enabling interrupts at all if MMU is enabled would mean that if another kind of interrupt is being handled, interrupts won't be re-enabled as well. Can't this be done more specific to the MMU handling case? Am I missing something?
@dcpleung any thoughts here?

Comment (dcpleung) [reply arch/x86/core/ia32/excstub.S:133]: I guess a better place to do this would be in the page fault handler where it disables interrupts until safe to re-enable again?

**Comment by andyross** [comment]
What's the circumstance where this is actually needed?  We aren't linux and we don't have lazy vma's that need to be populated at arbitrary times.  If you have an ISR that interrupts an exception handler, surely it's reasonable to demand that ISR not hit a page fault.  Why would it?  ISRs don't touch thread memory more or less by definition, and if they do it's under the control of a driver (e.g. a dma buffer) that can be held responsible for making sure things are mapped.

Short version: this looks like a workaround to a bug somewhere else.

File: arch/x86/core/ia32/excstub.S (hunk 1 above)
Comment (kwd-doodling) [reply arch/x86/core/ia32/excstub.S:133]: no existing good way to know if current exception is page fault exception.
@dcpleung Your way still has a short window with interrupt enabled, which is still not safe.
And I doubt it's really necessary to enable interrupt during an exception handling. Fire an interrupt that could be any type during an exception handling, the ISR may wake up another thread with higher priority, and swap code after ISR will switch to that thread to run, leaving half-way exception handling.

**Comment by kwd-doodling** [comment]
> What's the circumstance where this is actually needed? We aren't linux and we don't have lazy vma's that need to be populated at arbitrary times. If you have an ISR that interrupts an exception handler, surely it's reasonable to demand that ISR not hit a page fault. Why would it? ISRs don't touch thread memory more or less by definition, and if they do it's under the control of a driver (e.g. a dma buffer) that can be held responsible for making sure things are mapped.
> 
> Short version: this looks like a workaround to a bug somewhere else.

Need to go through all ISRs and make sure everything they call pinned, all kernel APIs, callback functions registered by user. this is buggy.
Any more, if ISR wakes up a thread with higher priority, swap function called at the end of ISR common code can swap CPU to that   thread and run thread code first.

**Review by andyross — CHANGES_REQUESTED** [review]
> Need to go through all ISRs and make sure everything they call pinned, all kernel APIs, callback functions registered by user. this is buggy.

Which is a fair point, but better addressed by (1) fixing the bugs and probably (2) coming up with a validation/assert framework to prevent that.  The cost here is really **extremely** high for an OS that claims to be "real time", this isn't the right way to write defensive code.  I think I'm a -1 unless there's a clear situation where code can't work correctly without this.

**Comment by kwd-doodling** [comment]
> > Need to go through all ISRs and make sure everything they call pinned, all kernel APIs, callback functions registered by user. this is buggy.
> 
> Which is a fair point, but better addressed by (1) fixing the bugs and probably (2) coming up with a validation/assert framework to prevent that. The cost here is really **extremely** high for an OS that claims to be "real time", this isn't the right way to write defensive code. I think I'm a -1 unless there's a clear situation where code can't work correctly without this.

I understand your concern about real-time, but if you have MMU/paging enabled that means you already accept a degree of real-time lost, assuming paging in/out time is quick and controllable. 
from the very beginning, ISH went the way you mentioned, but as project went several performance issues were reported when running sensing usages. To fix them, we have to reduce things pinned as much as possible.

I'm not familiar with how Linux deal with this, but from googling that Linux enables interrupt for some kinds of exceptions which allow re-entrance. It seems to always enable interrupt is not a right way.
